### PR TITLE
feat: add runtime version to framework integrations

### PR DESCRIPTION
**Pull request by tomlongridge**
## Goal

Adds version information for each integration with bottle being a conditional import
within wsgi. Also added a bottle example project.

### Added

- Bottle example project

### Changed

- celery/__init__.py - to add celery.__version__
- django/__init__.py - to add django.__version__
- flask/__init__.py - to add flask.__version__
- tornado/__init__.py - to add tornado.version
- wsgi/middleware.py - to add bottle.__version__ (if available)

## Tests

Re-ran example projects to ensure deviceData was populated as expected.

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [ ] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: example/bottle/server.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    """Deliberately raises an unhandled error and crash the app.
+    """
+    raise(Exception('SomethingBad'))
+
+@route('/handled')
+def handled():
+    try:
+        x = 1/0
+    except ZeroDivisionError:
+        bugsnag.notify(ZeroDivisionError('Bottle demo: To infinity... and beyond!'))
```
Comment (tobyhs) [inline example/bottle/server.py]: Something seems off here. You catch the `ZeroDivisionError` but don't use it.

It seems like this code do one of the following:
1) have `except ZeroDivisionError as e:` on the previous line, then call `bugsnag.notify(e)`
2) Avoid the `try`+`except` and just call `bugsnag.notify(ZeroDivisionError(...))`

File: bugsnag/wsgi/middleware.py [hunk 2]
```diff
@@ -37,6 +42,8 @@ def __init__(self, application, environ, start_response):
         self.environ = environ
 
         bugsnag.configure_request(wsgi_environ=self.environ)
+        if bottle_present:
+            bugsnag.configure().runtime_versions['bottle'] = bottle.__version__
```
Comment (tobyhs) [inline bugsnag/wsgi/middleware.py]: Should this be moved to `BugsnagMiddleware.__init__` instead?

It seems like setting the bottle version only needs to happen once, but I think `WrappedWSGIApp.__init__` is called for every request.

File: bugsnag/django/__init__.py [hunk 3]
```diff
@@ -69,6 +70,8 @@ def configure():
     middleware = bugsnag.configure().internal_middleware
     middleware.before_notify(add_django_request_to_notification)
 
+    bugsnag.configure().runtime_versions['django'] = django.__version__
```
Comment (tobyhs) [inline bugsnag/django/__init__.py]: Should there be something in tests/integrations/test_django_1.py that checks for the presence of this entry? (similar question applies to the Flask integration too)

File: example/bottle/server.py (hunk 1 above)
Comment (tomlongridge) [reply example/bottle/server.py]: I have just lifted the functions from the `flask` example, but I take your point. I've amended both to match the `plain` example.

File: bugsnag/wsgi/middleware.py (hunk 2 above)
Comment (tomlongridge) [reply bugsnag/wsgi/middleware.py]: Good spot - thanks, I hadn't noticed that.

File: bugsnag/celery/__init__.py [hunk 4]
```diff
@@ -28,4 +29,5 @@ def connect_failure_handler():
     Connect the bugsnag failure_handler to the Celery
     task_failure signal
     """
+    bugsnag.configure().runtime_versions['celery'] = celery.__version__
```
Comment (kylef) [inline bugsnag/celery/__init__.py]: I think this change is missing test coverage.

File: bugsnag/tornado/__init__.py [hunk 5]
```diff
@@ -38,6 +39,7 @@ def _handle_request_exception(self, exc):
         RequestHandler._handle_request_exception(self, exc)
 
     def prepare(self):
+        bugsnag.configure().runtime_versions['tornado'] = tornado.version
```
Comment (kylef) [inline bugsnag/tornado/__init__.py]: I think this change is missing test coverage.

File: example/django/requirements.txt [hunk 6]
```diff
@@ -1,2 +1,2 @@
-Django>=2.0.1
+Django==2.0.1
```
Comment (kylef) [inline example/django/requirements.txt]: Why would we want to pin this example to an outdated version patch of Django 2? This particular version of Django has known vulnerabilities registered as CVEs.

File: example/flask/server.py [hunk 7]
```diff
@@ -100,16 +100,15 @@ def crashcallback():
...
 
 @app.route('/handled')
-def notify():
+def handle_zero_div():
     """Deliberately triggers a handled exception, and reports it to Bugsnag.
     """
     try:
         x = 1/0
-    except ZeroDivisionError:
-        bugsnag.notify(ZeroDivisionError('Flask demo: To infinity... and beyond!'))
```
Comment (kylef) [inline example/flask/server.py]: This change makes this particular demo inconsistent with the rest of the examples (Django 2, Django 1, celery). What's the justification for that (and how is that related to the runtime version)?

File: tests/integrations/test_flask.py [hunk 8]
```diff
@@ -211,3 +212,21 @@ def test():
...
+        handle_exceptions(app)
+        app.test_client().get('/hello')
+
+        self.assertEqual(len(self.server.received), 1)
+
+        payload = self.server.received[0]['json_body']
+        device_data = payload['events'][0]['device']
+
+        self.assertEqual(device_data['runtimeVersions']['flask'],
+                         flask.__version__)
```
Comment (kylef) [inline tests/integrations/test_flask.py]: Perhaps it would be a good idea to test a bit more details about the version set here, the test lacks protection against changes in flask regarding the `__version__` and gives a false sense of security.

For example, if in the next major version of Flask the API of `__version__` changes to be an array of numbers or an object (example: `__version__ = {'major': 18, 'minor': 0, 'patch': 0}`) this test would continue to pass after upgrading flask. That means anyone in the future going to ensure that the integration works with a new version of Flask (the same applies to the other integrations) may wrongly expect that the code is compatible with the newer version. Which isn't correct, it would unexpectedly change the format of the version sent to the server.

It may be a good idea to also test that `device_data['runtimeVersions']['flask']` matches a regex like `\d+\.\d+\.\d+` which ensures it is a string of the expected format?

File: bugsnag/wsgi/middleware.py [hunk 9]
```diff
@@ -1,8 +1,17 @@
 from webob import Request
+import sys
 
 import bugsnag
 from bugsnag.wsgi import request_path
 
+bottle_present = False
+if 'bottle' in sys.modules:
```
Comment (kylef) [inline bugsnag/wsgi/middleware.py]: Perhaps a little more idomatic to use:

```python
try:
    import bottle
except ImportError:
    bottle = None

if bottle:
    bugsnag.configure().runtime_versions['bottle'] = bottle.__version__
```

File: bugsnag/wsgi/middleware.py [hunk 10]
```diff
@@ -1,8 +1,17 @@
...
+    bottle_present = True
+    try:
+        import bottle
```
Comment (kylef) [inline bugsnag/wsgi/middleware.py]: I wonder if this could give false positives, just because `bottle` is importable does not mean the code base is using it.

Bottle could be installed as a global dependency on a server (perhaps on a shared web hosted, such as `apt-get install python-bottle`), there may be other Python applications using bugsnag-python running on that machine which would incorrectly report they are using bottle of a given version. I am wondering more about shared hosting providers.

File: example/bottle/README.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+This Bottle application demonstrates how to use Bugsnag with the Bottle web framework for Python.
+
+## Setup
+
+Try this out with [your own Bugsnag account](https://app.bugsnag.com/user/new)! You'll be able to see how the errors are reported in the dashboard, how breadcrumbs are left, how errors are grouped and how they relate to the original source.
+
+To get set up, follow the instructions below. Don't forget to replace the placeholder API token with your own!
+
+1. Clone the repo and `cd` into this directory:
+    ```sh
```
Comment (kylef) [inline example/bottle/README.md]:
```suggestion
    ```shell
```

You're using `shell` everywhere else in this file.

File: example/bottle/README.md [hunk 12]
```diff
@@ -0,0 +1,33 @@
+# Bugsnag Bottle demo
```
Comment (kylef) [inline example/bottle/README.md]: This new integration example does not seem to have any tests.

File: example/bottle/index.html [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+    <div id="python">
+        <a href="/crash" style="text-decoration:none;">
+          <button type="button">UNHANDLED ERROR</button>
+        </a>
+        <br>
+        <a href="/handled" style="text-decoration:none;">
+          <button type="button">HANDLED ERROR</button>
+        </a>
+    </div>
+<!-- review settings.py & views.py to see how bugsnag integrates with your app. -->
```
Comment (kylef) [inline example/bottle/index.html]: Should this be a comment? Perhaps it would be a good idea to make this an actual paragraph so readers of the rendered HTML page can view the insight.

File: example/bottle/server.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+@route('/')
+def index():
+    return static_file('index.html', root='.')
+
+@route('/crash')
+def crash():
+    """Deliberately raises an unhandled error and crash the app.
+    """
+    raise(Exception('SomethingBad'))
```
Comment (kylef) [inline example/bottle/server.py]: raise is not a function

```suggestion
    raise Exception('SomethingBad')
```

also flake8 is failing on this file:

```
$ flake8 example/bottle/server.py
example/bottle/server.py:10:1: E302 expected 2 blank lines, found 1
example/bottle/server.py:14:1: E302 expected 2 blank lines, found 1
example/bottle/server.py:20:1: E302 expected 2 blank lines, found 1
example/bottle/server.py:25:9: F841 local variable 'x' is assigned to but never used
example/bottle/server.py:25:14: E226 missing whitespace around arithmetic operator
example/bottle/server.py:29:80: E501 line too long (93 > 79 characters)
example/bottle/server.py:31:1: E305 expected 2 blank lines after class or function definition, found 1
```

File: tests/integrations/test_flask.py (hunk 8 above)
Comment (tomlongridge) [reply tests/integrations/test_flask.py]: Good point - I'll amend.

File: bugsnag/wsgi/middleware.py (hunk 10 above)
Comment (tobyhs) [reply bugsnag/wsgi/middleware.py]: I think that was what line 8 (`if 'bottle' in sys.modules`) is supposed to partially account for (if the user imported bottle, then they are likely using it).

File: example/flask/server.py (hunk 7 above)
Comment (tobyhs) [reply example/flask/server.py]: It wasn't related to runtime version directly; it was related to a previous comment in this pull request (https://github.com/bugsnag/bugsnag-python/pull/182#discussion_r294985483) for the new bottle example. I would have preferred that similar changes to different example apps be separate changes or pull requests, but from my previous jobs/companies, fighting every battle that I wanted was too mentally exhausting and lead to me being heavily criticized as "too picky" (and other similar things), so I had to pick my battles.

File: bugsnag/wsgi/middleware.py (hunk 10 above)
Comment (tomlongridge) [reply bugsnag/wsgi/middleware.py]: Indeed, I believe this will only trigger if bottle is imported elsewhere.

File: example/bottle/README.md (hunk 12 above)
Comment (tomlongridge) [reply example/bottle/README.md]: This isn't a new integration as such - it uses wsgi - I just added an extra example whilst I was testing the runtime versioning.

File: bugsnag/tornado/__init__.py (hunk 5 above)
Comment (tomlongridge) [reply bugsnag/tornado/__init__.py]: It is, so the change was manually tested using the example. I've raised an internal ticket to consider adding unit tests to this integration.

File: bugsnag/celery/__init__.py (hunk 4 above)
Comment (tomlongridge) [reply bugsnag/celery/__init__.py]: It is, so the change was manually tested using the example. I've raised an internal ticket to consider adding unit tests to this integration.

File: example/flask/server.py (hunk 7 above)
Comment (tomlongridge) [reply example/flask/server.py]: I agree with our comments @tobyhs on #182 - the examples are already a little inconsistent and break style rules, so I've raised a ticket to sort this out separately.
